precompiles/batch/src/lib.rs

```diff
@@ -60,20 +58,53 @@ pub fn log_subcall_failed(address: impl Into<H160>, index: usize) -> Log {
 }
 
 /// Batch precompile.
-pub struct BatchPrecompile<Runtime>(PhantomData<Runtime>);
+#[derive(Debug, Clone)]
+pub struct BatchPrecompile<Runtime> {
+	/// Recursion level that will be increased when entering the precompile,
+	/// and decreased when exiting it. RefCell allows to mutate this value
+	/// while only having a shared reference. This is safe as the execution
+	/// is single-threaded and we don't hold the reference while doing subcalls.
+	recursion_level: RefCell<u8>,
+	/// Number of time the precompile can be called recursively.
+	/// 1 = can be called but subcalls cannot call again this precompile
+	/// ...
+	max_recursion_level: u8,
+	_phantom: PhantomData<Runtime>,
+}
 
-impl<Runtime> Precompile for BatchPrecompile<Runtime>
+impl<Runtime> StatefulPrecompile for BatchPrecompile<Runtime>
 where
 	Runtime: pallet_evm::Config,
 {
-	fn execute(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+	fn execute(&self, handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
 		let selector = handle.read_selector()?;
 
 		// No funds are transfered to the precompile address.
 		// Transfers will directly be made on the behalf of the user by the precompile.
 		handle.check_function_modifier(FunctionModifier::NonPayable)?;
 
+		match self.recursion_level.try_borrow_mut() {
+			Ok(mut recursion_level) => {
+				if *recursion_level >= self.max_recursion_level {
+					return Err(revert("Too much calls to Batch precompiles are nested"));
```

Comment:
I think here we should reset the `recursion_level` to 0. I'm thinking i.e. when a block is built, a single instance of the runtime is created. If some of the batch reaches the `max_recursion_level`, the evm reverts but the reference still maxed out. We'd also need a test for it.

Reply:
A new PrecompileSet instance is created at the start of every transaction, thus getting a fresh 0 level.

Reply:
Also we don't want to reset to 0, as if the upper batch is a batchSome it will continue trying executing more subcalls. This revert doesn't mean the whole transaction will revert, only the subcall to this too deeply nested batch.

Reply:
Great, thank you for the explanation.